R/utils_render_rtf_redo.R

```diff
@@ -0,0 +1,1283 @@
+      paste0(
+        rtf_key("red", rgb[1]),
+        rtf_key("green", rgb[2]),
+        rtf_key("blue", rgb[3])
+      )
+    )
+
+  class(rtf_color) <- "rtf_color"
+  rtf_color
+}
+
+rtf_header <- function(..., .charset = "ansi", .ansi_code_page = 1252) {
+
+  # Requires the font table (<fonttbl>)
+
+  # Optionally can use the:
+  #   [1] file table (<filetbl>)
+  #   [2] color table (<colortbl>)
+  #   [3] stylesheet (<stylesheet>)
+  #   [4] list tables (<listtables>)
+  #   [5] revision table (<revtbl>)
+
+  dots_list <- list(...)
+
+  additional_items <- c()
+
+  # TODO: Validate `dots_list` for object types
+
+  rtf_rtf_key <- rtf_key("rtf")
+  charset_key <- rtf_key(.charset)
+  ansi_code_page_key <- rtf_key("ansicpg", .ansi_code_page)
+
+  # Get a vector of the different object types
+  object_types <- vapply(dots_list, class, FUN.VALUE = character(1), USE.NAMES = FALSE)
+
+  # TODO: Validate the `object_types` vector
+
+  # Extract the `font_table` object
+  font_table <- unlist(dots_list[object_types == "rtf_fonttbl"])
+
+  if ("rtf_colortbl" %in% object_types) {
+    additional_items <- c(additional_items, unlist(dots_list[object_types == "rtf_colortbl"]))
+  }
+
+  # Arrange the keywords of the RTF header
+  rtf_header <-
+    list(
+      rtf_header = paste0(
+        rtf_rtf_key,
+        charset_key,
+        ansi_code_page_key,
+        font_table,
+        additional_items,
+        "\n"
+      )
+    )
+
+  class(rtf_header) <- c("rtf_header", "rtf")
+  rtf_header
+}
+
+rtf_document <- function(.file, ...) {
+
+  rtf_document <- list(...)
+
+  class(rtf_document) <- c("rtf_document", "rtf")
+  rtf_document
+}
+
+rtf_file <- function(header = NULL, document = NULL) {
+
+  if (is.null(header) && is.null(document)) {
+    header <- document <- ""
+  } else if (is.null(header) && !is.null(document)) {
+
+    # If `header` is NULL then generate the header based on the
+    # `document` content (and also modify the document)
+
+    # Scan for hexadecimal colors in the document; generate
+    # a <colortbl> object
+    matched_colors <-
+      unique(unlist(stringr::str_extract_all(document, "<<#[0-9a-fA-F]{6}>>")))
+
+    if (length(matched_colors) > 0) {
+
+      colortbl <-
+        rtf_colortbl(.hexadecimal_colors = gsub("<<|>>", "", matched_colors))
+
+      for (i in seq_along(matched_colors)) {
+        document <- gsub(matched_colors[i], as.character(i), document)
+      }
+    } else {
+      colortbl <- rtf_colortbl(.hexadecimal_colors = "#FFFFFF")
+    }
+
+    # Scan for hexadecimal colors in the document; generate
+    # a <fonttbl> object
+    matched_fonts <-
+      unique(unlist(stringr::str_extract_all(document, "<<FONT:.*?>>")))
+
+    if (length(matched_fonts) > 0) {
+
+      fonttbl <-
+        rtf_fonttbl(.font_names = c("Calibri", gsub("<<FONT:|>>", "", matched_fonts)), .default_n = 1)
+
+      for (i in seq_along(matched_fonts)) {
+        document <- gsub(matched_fonts[i], as.character(i), document)
```

Comment:
`fixed = TRUE`

Reply:
Now changed (and in the similar instance above).